Handle "np.asfarray was removed in the NumPy 2.0 release"

Example log: https://github.com/astropy/astropy/actions/runs/5989440429/job/16245628701

Need to update the `astropy/units/quantity_helper/function_helpers.py` module.

```
astropy/units/__init__.py:23: in <module>
    from .quantity import *
astropy/units/quantity.py:35: in <module>
    from .quantity_helper import can_have_arbitrary_unit, check_output, converters_and_unit
astropy/units/quantity_helper/__init__.py:16: in <module>
    from . import erfa, function_helpers, helpers, scipy_special
astropy/units/quantity_helper/function_helpers.py:185: in <module>
    np.copy, np.asfarray, np.real_if_close, np.sort_complex, np.resize,
numpy/__init__.py:351: in __getattr__
    raise AttributeError(
E   AttributeError: `np.asfarray` was removed in the NumPy 2.0 release. Use `np.asarray` with a proper dtype instead.
```

--- a/astropy/io/ascii/cparser.pyx
+++ b/astropy/io/ascii/cparser.pyx
@@ -733,7 +733,7 @@ cdef class CParser:
         if nrows != -1:
             num_rows = nrows
 
-        cdef np.ndarray col = np.empty(num_rows, dtype=np.float_)
+        cdef np.ndarray col = np.empty(num_rows, dtype=np.float64)
         cdef double converted
         cdef int row = 0
         cdef double *data = <double *> col.data

--- a/astropy/io/ascii/tests/test_c_reader.py
+++ b/astropy/io/ascii/tests/test_c_reader.py
@@ -1244,7 +1244,7 @@ def test_data_out_of_range(parallel, fast_reader, guess):
         fast_reader["parallel"] = parallel
         if fast_reader.get("use_fast_converter"):
             rtol = 1.0e-15
-        elif sys.maxsize < 2**32:
+        elif (sys.maxsize < 2**32) or (sys.platform == "win32"):
             # On 32bit the standard C parser (strtod) returns strings for these
             pytest.xfail("C parser cannot handle float64 on 32bit systems")
 
@@ -1359,7 +1359,7 @@ def test_data_at_range_limit(parallel, fast_reader, guess):
         fast_reader["parallel"] = parallel
         if fast_reader.get("use_fast_converter"):
             rtol = 1.0e-15
-        elif sys.maxsize < 2**32:
+        elif (sys.maxsize < 2**32) or (sys.platform == "win32"):
             # On 32bit the standard C parser (strtod) returns strings for these
             pytest.xfail("C parser cannot handle float64 on 32bit systems")
 

--- a/astropy/io/ascii/tests/test_c_reader.py
+++ b/astropy/io/ascii/tests/test_c_reader.py
@@ -4,6 +4,7 @@ import functools
 import io
 import os
 import re
+import sys
 from contextlib import nullcontext
 from io import BytesIO
 from textwrap import dedent
@@ -1243,7 +1244,7 @@ def test_data_out_of_range(parallel, fast_reader, guess):
         fast_reader["parallel"] = parallel
         if fast_reader.get("use_fast_converter"):
             rtol = 1.0e-15
-        elif np.iinfo(np.int_).dtype == np.dtype(np.int32):
+        elif sys.maxsize < 2**32:
             # On 32bit the standard C parser (strtod) returns strings for these
             pytest.xfail("C parser cannot handle float64 on 32bit systems")
 
@@ -1358,7 +1359,7 @@ def test_data_at_range_limit(parallel, fast_reader, guess):
         fast_reader["parallel"] = parallel
         if fast_reader.get("use_fast_converter"):
             rtol = 1.0e-15
-        elif np.iinfo(np.int_).dtype == np.dtype(np.int32):
+        elif sys.maxsize < 2**32:
             # On 32bit the standard C parser (strtod) returns strings for these
             pytest.xfail("C parser cannot handle float64 on 32bit systems")
 

--- a/astropy/io/misc/tests/test_yaml.py
+++ b/astropy/io/misc/tests/test_yaml.py
@@ -41,7 +41,6 @@ from astropy.time import Time
         2.0,
         np.float64(),
         3 + 4j,
-        np.complex_(3 + 4j),
         np.complex64(3 + 4j),
         np.complex128(1.0 - 2**-52 + 1j * (1.0 - 2**-52)),
     ],

--- a/astropy/io/misc/yaml.py
+++ b/astropy/io/misc/yaml.py
@@ -252,7 +252,6 @@ AstropyDumper.add_representer(SerializedColumn, _serialized_column_representer)
 # Numpy dtypes
 AstropyDumper.add_representer(np.bool_, yaml.representer.SafeRepresenter.represent_bool)
 for np_type in [
-    np.int_,
     np.intc,
     np.intp,
     np.int8,
@@ -267,11 +266,11 @@ for np_type in [
     AstropyDumper.add_representer(
         np_type, yaml.representer.SafeRepresenter.represent_int
     )
-for np_type in [np.float_, np.float16, np.float32, np.float64, np.longdouble]:
+for np_type in [np.float16, np.float32, np.float64, np.longdouble]:
     AstropyDumper.add_representer(
         np_type, yaml.representer.SafeRepresenter.represent_float
     )
-for np_type in [np.complex_, complex, np.complex64, np.complex128]:
+for np_type in [complex, np.complex64, np.complex128]:
     AstropyDumper.add_representer(np_type, _complex_representer)
 
 AstropyLoader.add_constructor("tag:yaml.org,2002:python/complex", _complex_constructor)

--- a/astropy/io/votable/converters.py
+++ b/astropy/io/votable/converters.py
@@ -4,7 +4,6 @@ This module handles the conversion of various VOTABLE datatypes
 to/from TABLEDATA_ and BINARY_ formats.
 """
 
-
 # STDLIB
 import re
 import sys
@@ -1362,11 +1361,9 @@ numpy_dtype_to_field_mapping = {
     np.int64().dtype.num: "long",
     np.complex64().dtype.num: "floatComplex",
     np.complex128().dtype.num: "doubleComplex",
-    np.unicode_().dtype.num: "unicodeChar",
+    np.str_().dtype.num: "unicodeChar",
+    np.bytes_().dtype.num: "char",
 }
-
-
-numpy_dtype_to_field_mapping[np.bytes_().dtype.num] = "char"
 
 
 def _all_matching_dtype(column):

--- a/astropy/io/votable/tests/test_vo.py
+++ b/astropy/io/votable/tests/test_vo.py
@@ -11,16 +11,14 @@ import pathlib
 import sys
 from unittest import mock
 
-import numpy as np
-
 # THIRD-PARTY
+import numpy as np
 import pytest
 from numpy.testing import assert_array_equal
 
+# LOCAL
 from astropy.io.votable import tree
 from astropy.io.votable.exceptions import W39, VOTableSpecError, VOWarning
-
-# LOCAL
 from astropy.io.votable.table import parse, parse_single_table, validate
 from astropy.io.votable.xmlutil import validate_schema
 from astropy.utils.data import get_pkg_data_filename, get_pkg_data_filenames
@@ -298,7 +296,7 @@ class TestParse:
         )
 
     def test_fixed_string_test(self):
-        assert issubclass(self.array["string_test_2"].dtype.type, np.unicode_)
+        assert issubclass(self.array["string_test_2"].dtype.type, np.str_)
         assert_array_equal(
             self.array["string_test_2"], ["Fixed stri", "0123456789", "XXXX", "", ""]
         )
@@ -311,7 +309,7 @@ class TestParse:
         )
 
     def test_fixed_unicode_test(self):
-        assert issubclass(self.array["fixed_unicode_test"].dtype.type, np.unicode_)
+        assert issubclass(self.array["fixed_unicode_test"].dtype.type, np.str_)
         assert_array_equal(
             self.array["fixed_unicode_test"],
             ["Ceçi n'est", "வணக்கம்", "0123456789", "", ""],

--- a/astropy/modeling/tests/test_constraints.py
+++ b/astropy/modeling/tests/test_constraints.py
@@ -15,7 +15,6 @@ from astropy.modeling import fitting, models
 from astropy.modeling.core import Fittable1DModel
 from astropy.modeling.parameters import Parameter
 from astropy.utils import minversion
-from astropy.utils.compat.numpycompat import NUMPY_LT_2_0
 from astropy.utils.compat.optional_deps import HAS_SCIPY
 from astropy.utils.exceptions import AstropyUserWarning
 
@@ -184,6 +183,7 @@ class TestBounds:
         assert intercept + 10**-5 >= bounds["intercept"][0]
         assert intercept - 10**-5 <= bounds["intercept"][1]
 
+    @pytest.mark.filterwarnings("ignore:The fit may be unsuccessful")
     @pytest.mark.parametrize("fitter", fitters)
     def test_bounds_gauss2d_lsq(self, fitter):
         fitter = fitter()
@@ -204,21 +204,12 @@ class TestBounds:
             theta=0.5,
             bounds=bounds,
         )
-        if isinstance(fitter, (fitting.LevMarLSQFitter, fitting.DogBoxLSQFitter)):
-            with pytest.warns(AstropyUserWarning, match="The fit may be unsuccessful"):
-                model = fitter(gauss, X, Y, self.data)
+        if isinstance(fitter, fitting.TRFLSQFitter):
+            ctx = np.errstate(invalid="ignore", divide="ignore")
         else:
-            ctx2 = nullcontext()
-            if isinstance(fitter, fitting.TRFLSQFitter):
-                ctx = np.errstate(invalid="ignore", divide="ignore")
-                if not NUMPY_LT_2_0 or not SCIPY_LT_1_11_2:
-                    ctx2 = pytest.warns(
-                        AstropyUserWarning, match="The fit may be unsuccessful"
-                    )
-            else:
-                ctx = nullcontext()
-            with ctx, ctx2:
-                model = fitter(gauss, X, Y, self.data)
+            ctx = nullcontext()
+        with ctx:
+            model = fitter(gauss, X, Y, self.data)
         x_mean = model.x_mean.value
         y_mean = model.y_mean.value
         x_stddev = model.x_stddev.value

--- a/astropy/table/meta.py
+++ b/astropy/table/meta.py
@@ -193,7 +193,7 @@ def _get_variable_length_array_shape(col):
         """Local conversion error used below."""
 
     # Numpy types supported as variable-length arrays
-    np_classes = (np.floating, np.integer, np.bool_, np.unicode_)
+    np_classes = (np.floating, np.integer, np.bool_, np.str_)
 
     try:
         if len(col) == 0 or not all(isinstance(val, np.ndarray) for val in col):

--- a/astropy/table/tests/test_column.py
+++ b/astropy/table/tests/test_column.py
@@ -268,7 +268,7 @@ class TestColumn:
         Tests for #3095, which forces integer item access to always return a plain
         ndarray or MaskedArray, even in the case of a multi-dim column.
         """
-        integer_types = (int, np.int_)
+        integer_types = (int, np.int32, np.int64)
 
         for int_type in integer_types:
             c = Column([[1, 2], [3, 4]])

--- a/astropy/table/tests/test_row.py
+++ b/astropy/table/tests/test_row.py
@@ -348,16 +348,16 @@ def test_uint_indexing():
     that printing such a row works.
 
     This is non-trivial: adding a signed and unsigned
-    integer in numpy results in a float, which is an
+    64 bit integer in numpy results in a float, which is an
     invalid slice index.
 
     Regression test for gh-7464.
     """
     t = table.Table([[1.0, 2.0, 3.0]], names="a")
     assert t["a"][1] == 2.0
-    assert t["a"][np.int_(1)] == 2.0
-    assert t["a"][np.uint(1)] == 2.0
-    assert t[np.uint(1)]["a"] == 2.0
+    assert t["a"][np.int64(1)] == 2.0
+    assert t["a"][np.uint64(1)] == 2.0
+    assert t[np.uint64(1)]["a"] == 2.0
 
     trepr = [
         "<Row index=1>",
@@ -368,8 +368,8 @@ def test_uint_indexing():
     ]
 
     assert repr(t[1]).splitlines() == trepr
-    assert repr(t[np.int_(1)]).splitlines() == trepr
-    assert repr(t[np.uint(1)]).splitlines() == trepr
+    assert repr(t[np.int64(1)]).splitlines() == trepr
+    assert repr(t[np.uint64(1)]).splitlines() == trepr
 
 
 def test_row_get():

--- a/astropy/time/formats.py
+++ b/astropy/time/formats.py
@@ -2177,11 +2177,11 @@ class TimeDeltaDatetime(TimeDeltaFormat, TimeUnique):
 
 def _validate_jd_for_storage(jd):
     if isinstance(jd, (float, int)):
-        return np.array(jd, dtype=np.float_)
+        return np.array(jd, dtype=float)
     if isinstance(jd, np.generic) and (
         jd.dtype.kind == "f" and jd.dtype.itemsize <= 8 or jd.dtype.kind in "iu"
     ):
-        return np.array(jd, dtype=np.float_)
+        return np.array(jd, dtype=float)
     elif isinstance(jd, np.ndarray) and jd.dtype.kind == "f" and jd.dtype.itemsize == 8:
         return jd
     else:

--- a/astropy/units/quantity_helper/function_helpers.py
+++ b/astropy/units/quantity_helper/function_helpers.py
@@ -442,7 +442,7 @@ def select(condlist, choicelist, default=0):
 def piecewise(x, condlist, funclist, *args, **kw):
     from astropy.units import Quantity
 
-    # Copied implementation from numpy.lib.function_base.piecewise,
+    # Copied implementation from numpy.lib._function_base_impl.piecewise,
     # taking care of units of function outputs.
     n2 = len(funclist)
     # undocumented: single condition is promoted to a list of one condition

--- a/astropy/units/quantity_helper/function_helpers.py
+++ b/astropy/units/quantity_helper/function_helpers.py
@@ -182,13 +182,13 @@ dispatched_function = FunctionAssigner(DISPATCHED_FUNCTIONS)
 # fmt: off
 @function_helper(
     helps={
-        np.copy, np.asfarray, np.real_if_close, np.sort_complex, np.resize,
+        np.copy, np.real_if_close, np.sort_complex, np.resize,
         np.fft.fft, np.fft.ifft, np.fft.rfft, np.fft.irfft,
         np.fft.fft2, np.fft.ifft2, np.fft.rfft2, np.fft.irfft2,
         np.fft.fftn, np.fft.ifftn, np.fft.rfftn, np.fft.irfftn,
         np.fft.hfft, np.fft.ihfft,
         np.linalg.eigvals, np.linalg.eigvalsh,
-    }
+    } | ({np.asfarray} if NUMPY_LT_2_0 else set())
 )
 # fmt: on
 def invariant_a_helper(a, *args, **kwargs):

--- a/astropy/units/tests/test_quantity_non_ufuncs.py
+++ b/astropy/units/tests/test_quantity_non_ufuncs.py
@@ -17,7 +17,12 @@ from astropy.units.quantity_helper.function_helpers import (
     TBD_FUNCTIONS,
     UNSUPPORTED_FUNCTIONS,
 )
-from astropy.utils.compat import NUMPY_LT_1_23, NUMPY_LT_1_24, NUMPY_LT_1_25
+from astropy.utils.compat import (
+    NUMPY_LT_1_23,
+    NUMPY_LT_1_24,
+    NUMPY_LT_1_25,
+    NUMPY_LT_2_0,
+)
 
 needs_array_function = pytest.mark.xfail(
     not ARRAY_FUNCTION_ENABLED, reason="Needs __array_function__ support"
@@ -296,6 +301,7 @@ class TestCopyAndCreation(InvariantUnitTestSetup):
         copy = np.copy(a=self.q)
         assert_array_equal(copy, self.q)
 
+    @pytest.mark.skipif(not NUMPY_LT_2_0, reason="np.asfarray is removed in NumPy 2.0")
     @needs_array_function
     def test_asfarray(self):
         self.check(np.asfarray)

--- a/astropy/utils/masked/core.py
+++ b/astropy/utils/masked/core.py
@@ -19,6 +19,7 @@ import builtins
 
 import numpy as np
 
+from astropy.utils.compat import NUMPY_LT_2_0
 from astropy.utils.data_info import ParentDtypeInfo
 from astropy.utils.shapes import NDArrayShapeMethods
 
@@ -759,9 +760,17 @@ class MaskedNDArray(Masked, np.ndarray, base_cls=np.ndarray, data_cls=np.ndarray
             else:
                 # Parse signature with private numpy function. Note it
                 # cannot handle spaces in tuples, so remove those.
-                in_sig, out_sig = np.lib.function_base._parse_gufunc_signature(
-                    ufunc.signature.replace(" ", "")
-                )
+                if NUMPY_LT_2_0:
+                    in_sig, out_sig = np.lib.function_base._parse_gufunc_signature(
+                        ufunc.signature.replace(" ", "")
+                    )
+                else:
+                    (
+                        in_sig,
+                        out_sig,
+                    ) = np.lib._function_base_impl._parse_gufunc_signature(
+                        ufunc.signature.replace(" ", "")
+                    )
                 axis = kwargs.get("axis", -1)
                 keepdims = kwargs.get("keepdims", False)
                 in_masks = []

--- a/astropy/utils/masked/function_helpers.py
+++ b/astropy/utils/masked/function_helpers.py
@@ -1068,7 +1068,8 @@ def masked_nanfunc(nanfuncname):
     return nanfunc
 
 
-for nanfuncname in np.lib.nanfunctions.__all__:
+_nplibnanfunctions = np.lib.nanfunctions if NUMPY_LT_2_0 else np.lib._nanfunctions_impl
+for nanfuncname in _nplibnanfunctions.__all__:
     globals()[nanfuncname] = dispatched_function(
         masked_nanfunc(nanfuncname), helps=getattr(np, nanfuncname)
     )

--- a/astropy/utils/masked/function_helpers.py
+++ b/astropy/utils/masked/function_helpers.py
@@ -93,7 +93,7 @@ MASKED_SAFE_FUNCTIONS |= {
     np.isclose, np.allclose, np.flatnonzero, np.argwhere,
     # np.core.shape_base
     np.atleast_1d, np.atleast_2d, np.atleast_3d, np.stack, np.hstack, np.vstack,
-    # np.lib.function_base
+    # np.lib._function_base_impl
     np.average, np.diff, np.extract, np.meshgrid, np.trapz, np.gradient,
     # np.lib.index_tricks
     np.diag_indices_from, np.triu_indices_from, np.tril_indices_from,
@@ -108,7 +108,7 @@ MASKED_SAFE_FUNCTIONS |= {
     np.real_if_close, np.common_type,
     # np.lib.ufunclike
     np.fix, np.isneginf, np.isposinf,
-    # np.lib.function_base
+    # np.lib._function_base_impl
     np.angle, np.i0,
 }  # fmt: skip
 IGNORED_FUNCTIONS = {
@@ -121,7 +121,7 @@ IGNORED_FUNCTIONS = {
 IGNORED_FUNCTIONS |= {
     np.pad, np.searchsorted, np.digitize,
     np.is_busday, np.busday_count, np.busday_offset,
-    # numpy.lib.function_base
+    # numpy.lib._function_base_impl
     np.cov, np.corrcoef, np.trim_zeros,
     # numpy.core.numeric
     np.correlate, np.convolve,
@@ -439,7 +439,7 @@ if NUMPY_LT_2_0:
 
 @dispatched_function
 def sort_complex(a):
-    # Just a copy of function_base.sort_complex, to avoid the asarray.
+    # Just a copy of np.lib._function_base_impl.sort_complex, to avoid the asarray.
     b = a.copy()
     b.sort()
     if not issubclass(b.dtype.type, np.complexfloating):  # pragma: no cover
@@ -595,10 +595,14 @@ def median(a, axis=None, out=None, **kwargs):
         )
         return (r.reshape(k) if keepdims else r) if out is None else out
 
-    else:
+    elif NUMPY_LT_2_0:
         return np.lib.function_base._ureduce(
             a, func=_masked_median, axis=axis, out=out, **kwargs
         )
+
+    return np.lib._function_base_impl._ureduce(
+        a, func=_masked_median, axis=axis, out=out, **kwargs
+    )
 
 
 def _masked_quantile_1d(a, q, **kwargs):
@@ -608,7 +612,12 @@ def _masked_quantile_1d(a, q, **kwargs):
     """
     unmasked = a.unmasked[~a.mask]
     if unmasked.size:
-        result = np.lib.function_base._quantile_unchecked(unmasked, q, **kwargs)
+        if NUMPY_LT_2_0:
+            result = np.lib.function_base._quantile_unchecked(unmasked, q, **kwargs)
+        else:
+            result = np.lib._function_base_impl._quantile_unchecked(
+                unmasked, q, **kwargs
+            )
         return a.from_unmasked(result)
     else:
         return a.from_unmasked(np.zeros_like(a.unmasked, shape=q.shape), True)
@@ -641,7 +650,9 @@ def quantile(a, q, axis=None, out=None, **kwargs):
 
     a = Masked(a)
     q = np.asanyarray(q)
-    if not np.lib.function_base._quantile_is_valid(q):
+    if (NUMPY_LT_2_0 and not np.lib.function_base._quantile_is_valid(q)) or (
+        not NUMPY_LT_2_0 and not np.lib._function_base_impl._quantile_is_valid(q)
+    ):
         raise ValueError("Quantiles must be in the range [0, 1]")
 
     if NUMPY_LT_1_24:
@@ -650,10 +661,14 @@ def quantile(a, q, axis=None, out=None, **kwargs):
             a, func=_masked_quantile, q=q, axis=axis, out=out, **kwargs
         )
         return (r.reshape(q.shape + k) if keepdims else r) if out is None else out
-    else:
+    elif NUMPY_LT_2_0:
         return np.lib.function_base._ureduce(
             a, func=_masked_quantile, q=q, axis=axis, out=out, **kwargs
         )
+
+    return np.lib._function_base_impl._ureduce(
+        a, func=_masked_quantile, q=q, axis=axis, out=out, **kwargs
+    )
 
 
 @dispatched_function
@@ -761,7 +776,7 @@ def piecewise(x, condlist, funclist, *args, **kw):
     Any masks in ``condlist`` are ignored.
 
     """
-    # Copied implementation from numpy.lib.function_base.piecewise,
+    # Copied implementation from numpy.lib._function_base_impl.piecewise,
     # just to ensure output is Masked.
     n2 = len(funclist)
     # undocumented: single condition is promoted to a list of one condition

--- a/astropy/utils/masked/function_helpers.py
+++ b/astropy/utils/masked/function_helpers.py
@@ -103,7 +103,7 @@ MASKED_SAFE_FUNCTIONS |= {
     np.array_split, np.split, np.hsplit, np.vsplit, np.dsplit,
     np.expand_dims, np.apply_along_axis, np.kron, np.tile,
     np.take_along_axis, np.put_along_axis,
-    # np.lib.type_check (all but asfarray, nan_to_num)
+    # np.lib.type_check (all but nan_to_num)
     np.iscomplexobj, np.isrealobj, np.imag, np.isreal, np.real,
     np.real_if_close, np.common_type,
     # np.lib.ufunclike
@@ -206,7 +206,10 @@ def nan_to_num(x, copy=True, nan=0.0, posinf=None, neginf=None):
 # should be applied to the data and the mask.  They cannot all share the
 # same helper, because the first arguments have different names.
 @apply_to_both(
-    helps={np.copy, np.asfarray, np.resize, np.moveaxis, np.rollaxis, np.roll}
+    helps=(
+        {np.copy, np.resize, np.moveaxis, np.rollaxis, np.roll}
+        | ({np.asfarray} if NUMPY_LT_2_0 else set())
+    )
 )
 def masked_a_helper(a, *args, **kwargs):
     data, mask = _get_data_and_masks(a)

--- a/astropy/utils/masked/tests/test_function_helpers.py
+++ b/astropy/utils/masked/tests/test_function_helpers.py
@@ -8,7 +8,6 @@ in test_functions.
 TODO: finish full coverage (see also `~astropy.utils.masked.function_helpers`)
 - np.linalg
 - np.fft (is there any point?)
-- np.lib.nanfunctions
 
 """
 import inspect

--- a/astropy/utils/masked/tests/test_function_helpers.py
+++ b/astropy/utils/masked/tests/test_function_helpers.py
@@ -19,7 +19,12 @@ import pytest
 from numpy.testing import assert_array_equal
 
 from astropy.units.tests.test_quantity_non_ufuncs import get_wrapped_functions
-from astropy.utils.compat import NUMPY_LT_1_23, NUMPY_LT_1_24, NUMPY_LT_1_25
+from astropy.utils.compat import (
+    NUMPY_LT_1_23,
+    NUMPY_LT_1_24,
+    NUMPY_LT_1_25,
+    NUMPY_LT_2_0,
+)
 from astropy.utils.masked import Masked, MaskedNDArray
 from astropy.utils.masked.function_helpers import (
     APPLY_TO_BOTH_FUNCTIONS,
@@ -279,6 +284,7 @@ class TestCopyAndCreation(InvariantMaskTestSetup):
         copy = np.copy(a=self.ma)
         assert_array_equal(copy, self.ma)
 
+    @pytest.mark.skipif(not NUMPY_LT_2_0, reason="np.asfarray is removed in NumPy 2.0")
     def test_asfarray(self):
         self.check(np.asfarray)
         farray = np.asfarray(a=self.ma)

--- a/astropy/wcs/wcsapi/low_level_api.py
+++ b/astropy/wcs/wcsapi/low_level_api.py
@@ -128,7 +128,7 @@ class BaseLowLevelWCS(metaclass=abc.ABCMeta):
         else:
             pixel_arrays = pixel_arrays[::-1]
         array_indices = tuple(
-            np.asarray(np.floor(pixel + 0.5), dtype=np.int_) for pixel in pixel_arrays
+            np.asarray(np.floor(pixel + 0.5), dtype=int) for pixel in pixel_arrays
         )
         return array_indices[0] if self.pixel_n_dim == 1 else array_indices
 

--- a/tox.ini
+++ b/tox.ini
@@ -102,7 +102,7 @@ deps =
 # The following indicates which extras_require from setup.cfg will be installed
 extras =
     test: test
-    devdeps: test_all
+    devdeps: test
     recdeps: recommended
     recdeps: test_all
     alldeps: all

